### PR TITLE
Document DenseArrays and Dims

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -71,8 +71,27 @@ Union type of [`Vector{T}`](@ref) and [`Matrix{T}`](@ref).
 """
 const VecOrMat{T} = Union{Vector{T}, Matrix{T}}
 
+"""
+    DenseVector{T}
+
+One-dimensional dense array with elements of type `T`, often used to represent
+a mathematical vector. Alias for `DenseArray{T,1}`. The elements of a dense array
+are stored contiguously in memory, with no striding.
```
Comment (mbauman) [inline base/array.jl]: I get what you're trying to say here, but I think I'd refrain from using the phrase "no striding." A contiguous array still has `strides` — it's just that they're arranged such that they cover the memory contiguously with no overlaps or skips.

**Comment by mbauman** [comment]
Nice, thanks so much for tackling my massive list!

How about just documenting `DenseArray` with the meaty stuff once and then having `DenseVector` and `DenseMatrix` just be pointers to it — like you've already done for `DenseVecOrMat`?

**Comment by kshyatt** [comment]
Is `DenseArray` exported?

**Comment by mbauman** [comment]
Yup.  It looks like my script in #26919 missed it — perhaps because it's a builtin `Core` binding.

**Comment by kshyatt** [comment]
Better?

File: base/array.jl [hunk 2]
```diff
@@ -71,8 +71,36 @@ Union type of [`Vector{T}`](@ref) and [`Matrix{T}`](@ref).
...
+a mathematical vector, matrix, or higher dimensional tensor.
+The elements of a dense array are stored contiguously in memory.
+"""
+DenseArray
+
+"""
+    DenseVector{T}
+
+One-dimensional [`DenseArray`](@ref) with elements of type `T`, often used to represent
+a mathematical vector. Alias for `DenseArray{T,1}`.
```
Comment (fredrikekre) [inline base/array.jl]: The "often used to represent a mathematical vector" part seems unnecessary here too.

File: base/array.jl [hunk 3]
```diff
@@ -71,8 +71,36 @@ Union type of [`Vector{T}`](@ref) and [`Matrix{T}`](@ref).
...
+"""
 const DenseVector{T} = DenseArray{T,1}
+
+"""
+    DenseMatrix{T}
+
+Two-dimensional [`DenseArray`](@ref) with elements of type `T`, often used to represent
+a mathematical vector. Alias for `DenseArray{T,2}`.
```
Comment (fredrikekre) [inline base/array.jl]: Likewise here, and in any case change vector -> matrix.

File: base/indices.jl [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+"""
+    Dims{N}
+
+An `NTuple` of `N` `Int`s used to represent the dimensions
```
Comment (fredrikekre) [inline base/indices.jl]: Perhaps change "An `NTuple` of `N` `Int`s used to ..." to "An alias for `NTuple{N,Int}` used to ..."

File: base/array.jl [hunk 5]
```diff
@@ -71,8 +71,36 @@ Union type of [`Vector{T}`](@ref) and [`Matrix{T}`](@ref).
 """
 const VecOrMat{T} = Union{Vector{T}, Matrix{T}}
 
+"""
+    DenseArray{T, N} <: AbstractArray{T,N}
+
+`N`-dimensional dense array with elements of type `T`, often used to represent
+a mathematical vector, matrix, or higher dimensional tensor.
```
Comment (fredrikekre) [inline base/array.jl]: I don't think we need the
> often used to represent a mathematical vector, matrix, or higher dimensional tensor.

part, these are used just as often for other things, no?

Comment (waldyrious) [reply base/array.jl]: I think it's useful to have those terms in the documentation to help with searching cross-referencing.

Comment (mbauman) [reply base/array.jl]: Sure, but there's nothing special about `DenseArray` that makes it particularly better at being a mathematical vector/matrix/tensor.

**Comment by kshyatt** [comment]
Test fails seem unrelated to me - anyone confirm? This good to go?

**Comment by mbauman** [comment]
Wow, that was a rough go on CI.  The ccall failure in FreeBSD has been noted before. Mac was filewatching, Travis Linux 64 was the offset array failure (since fixed), and CircleCI has `JULIA_CPU_CORES` in its config.

**Pull request merged**